python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -791,3 +812,51 @@ def info_string(self):
         lm_summary = self.load_metrics.summary()
         autoscaler_summary = self.summary()
         return "\n" + format_info_string(lm_summary, autoscaler_summary)
+
+
+class NodeTracker:
+    """Map nodes to their corresponding logs.
+
+    We need to be a little careful here. At an given point in time, node_id <->
+    ip can be interchangeably use, but the node_id -> ip relation is not
+    bijective _across time_ since IP addresses can be reused. Therefore, we
+    should treat node_id as the only unique identifier.
+
+    """
+
+    def __init__(self, log_dir, process_runner):
+        self.log_dir = log_dir
+        self.process_runner = process_runner
+
+        # Mapping from node_id -> (ip, node type, stdout_path, process runner)
+        # TODO(Alex): In the spirit of statelessness, we should try to load
+        # this mapping from the filesystem. _Technically_ this tracker is only
+        # used for "recent" failures though, so remembering old nodes is a best
+        # effort, therefore it's already correct
+        self.node_mapping = {}
```

Comment:
Can we cap the number of failed nodes we track? Otherwise this could grow infinitely over time?

Reply:
I thought about this, but decided against since: 

1. This is already bounded by the size of the node provide tag cache
2. Along the lines of (1), we can just restart the autoscaler to clear both caches
3. Python doesn't have a built in LRU cache, so we'd have to either add the dependency or implement one.

Reply:
Could we keep a simple list and pop off the front on exceeding a limit?

Reply:
sure